File: examples/language-features/VB.NET/dotNetFramework/v4.8/Baseline/ExtensionMethods/ExtensionMethods.vb
Option Strict On

Imports System.Collections.Generic
Imports System.Runtime.CompilerServices

Namespace Baseline.ExtensionMethods
    ' An extension method is marked with the Extension attribute and must live
    ' in a Module. C# uses a `this` modifier on the first parameter instead;
    ' VB's marker is the attribute.
    Public Module StringExtensions
        <Extension()>
        Public Function IsNullOrBlank(value As String) As Boolean
            Return value Is Nothing OrElse value.Trim().Length = 0
        End Function

        <Extension()>
        Public Function Repeat(value As String, times As Integer) As String
            Dim result As String = String.Empty

            For i As Integer = 1 To times
                result &= value
            Next

            Return result
        End Function
    End Module

    Public Module EnumerableExtensions
        ' Extending an interface makes the method available on every
        ' implementing type.
        <Extension()>
        Public Function CountEven(values As IEnumerable(Of Integer)) As Integer
            Dim count As Integer = 0

            For Each value As Integer In values
                If value Mod 2 = 0 Then
                    count += 1
                End If
            Next

            Return count
        End Function
    End Module

    Public Module Usage
        Public Function CallAsInstanceMethod() As String
            Return "ab".Repeat(3)
        End Function

        ' The same call written as the plain invocation it compiles to.
        Public Function CallAsPlainMethod() As String
            Return StringExtensions.Repeat("ab", 3)
        End Function

        Public Function Evens(values As List(Of Integer)) As Integer
            Return values.CountEven()
        End Function
    End Module
End Namespace

File: examples/language-features/CSharp/dotnet/10/10.0/library/CSharp3/ExtensionMethods/ExtensionMethods.cs
using System.Collections.Generic;

namespace Net10_CSharp10_Library.CSharp3.ExtensionMethods
{
    // An extension method lives in a non-generic static class and marks its
    // first parameter with this.
    public static class StringExtensions
    {
        public static bool IsNullOrBlank(this string value)
        {
            return value == null || value.Trim().Length == 0;
        }

        public static string Repeat(this string value, int times)
        {
            string result = string.Empty;
            for (int i = 0; i < times; i++)
            {
                result += value;
            }

            return result;
        }
    }

    public static class EnumerableExtensions
    {
        // Extending an interface makes the method available on every type that
        // implements it.
        public static int CountEven(this IEnumerable<int> values)
        {
            int count = 0;
            foreach (int value in values)
            {
                if (value % 2 == 0)
                {
                    count++;
                }
            }

            return count;
        }
    }

    public class Usage
    {
        public static string CallAsInstanceMethod()
        {
            return "ab".Repeat(3);
        }

        // The same call written as the plain static invocation it compiles to.
        public static string CallAsStaticMethod()
        {
            return StringExtensions.Repeat("ab", 3);
        }

        public static bool BlankCheck()
        {
            return "  ".IsNullOrBlank();
        }

        public static int Evens(List<int> values)
        {
            return values.CountEven();
        }
    }
}
